src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs

```diff
@@ -0,0 +1,415 @@
+            {
+                var sb = new StringBuilder();
+                if (binaryData.Length > 0)
+                {
+                    sb.AppendFormat("Unable to convert from storage format GrainStateEntity.Data={0}", binaryData);
+                }
+                else if (!string.IsNullOrEmpty(stringData))
+                {
+                    sb.AppendFormat("Unable to convert from storage format GrainStateEntity.StringData={0}", stringData);
+                }
+                if (dataValue != null)
+                {
+                    sb.AppendFormat("Data Value={0} Type={1}", dataValue, dataValue.GetType());
+                }
+
+                Log.Error(0, sb.ToString(), exc);
+                throw new AggregateException(sb.ToString(), exc);
+            }
+
+            return dataValue;
+        }
+
+        /// <summary>
+        /// Serialize to DynamoDB format in either binary or JSON format.
+        /// </summary>
+        /// <param name="grainState">The grain state data to be serialized</param>
+        /// <param name="entity">The DynamoDB entity</param>
+        internal void ConvertToStorageFormat(object grainState, GrainStateRecord entity)
+        {
+            int dataSize;
+            if (useJsonFormat)
+            {
+                // http://james.newtonking.com/json/help/index.html?topic=html/T_Newtonsoft_Json_JsonConvert.htm
+                entity.StringData = JsonConvert.SerializeObject(grainState, jsonSettings);
+                dataSize = STRING_DATA_PROPERTY_NAME.Length + entity.StringData.Length;
+
+                if (Log.IsVerbose3) Log.Verbose3("Writing JSON data size = {0} for grain id = Partition={1} / Row={2}",
+                    dataSize, entity.PartitionKey, entity.RowKey);
+            }
+            else
+            {
+                // Convert to binary format
+                entity.Data = SerializationManager.SerializeToByteArray(grainState);
+                dataSize = BINARY_DATA_PROPERTY_NAME.Length + entity.Data.Length;
+
+                if (Log.IsVerbose3) Log.Verbose3("Writing binary data size = {0} for grain id = Partition={1} / Row={2}",
+                    dataSize, entity.PartitionKey, entity.RowKey);
+            }
+
+            var pkSize = PK_PROPERTY_NAME.Length + entity.PartitionKey.Length;
+            var rkSize = RK_PROPERTY_NAME.Length + entity.RowKey.Length;
+            var versionSize = ETAG_PROPERTY_NAME.Length + entity.ETag.ToString().Length;
+
+            if ((pkSize + rkSize + versionSize + dataSize) > MAX_DATA_SIZE)
+            {
+                var msg = string.Format("Data too large to write to DynamoDB table. Size={0} MaxSize={1}", dataSize, MAX_DATA_SIZE);
+                throw new ArgumentOutOfRangeException("GrainState.Size", msg);
+            }
+        }
+
+        private string GetKeyString(GrainReference grainReference)
+        {
+            var key = string.Format("{0}_{1}", serviceId, grainReference.ToKeyString());
```

Comment:
The implementation here and in similar places is the same as the Azure one. My knowledge of Azure Table or DynamoDB is close to nothing, so I don't know how similar they are. I think that before merging this, we must convince ourselves that following the Azure oriented implementation gives the best throughput in DynamoDB too.